packages/node/src/integrations/contextlines.ts

```diff
@@ -27,62 +21,210 @@ interface ContextLinesOptions {
+    function onStreamError(e: Error): void {
+      // Mark file path as failed to read and prevent multiple read attempts.
+      LRU_FILE_CONTENTS_FS_READ_FAILED.set(path, 1);
+      DEBUG_BUILD && logger.error(`Failed to read file: ${path}. Error: ${e}`);
+      lineReaded.close();
+      lineReaded.removeAllListeners();
+      resolve();
+    }
+
+    // We need to handle the error event to prevent the process from crashing in < Node 16
+    // https://github.com/nodejs/node/pull/31603
+    stream.on('error', onStreamError);
+    lineReaded.on('error', onStreamError);
+    lineReaded.on('close', resolve);
+
+    lineReaded.on('line', line => {
+      lineNumber++;
+      if (lineNumber < rangeStart) return;
+
+      // !Warning: Mutates the cache value
+      output[lineNumber] = line;
+
+      if (lineNumber >= rangeEnd) {
+        if (currentRangeIndex === ranges.length - 1) {
+          // We need to close the file stream and remove listeners, else the reader will continue to run our listener;
+          lineReaded.close();
+          lineReaded.removeAllListeners();
+          return;
+        }
+        currentRangeIndex++;
+        rangeStart = ranges[currentRangeIndex][0];
+        rangeEnd = ranges[currentRangeIndex][1];
+      }
+    });
+  });
+}
 
 async function addSourceContext(event: Event, contextLines: number): Promise<Event> {
   // keep a lookup map of which files we've already enqueued to read,
   // so we don't enqueue the same file multiple times which would cause multiple i/o reads
-  const enqueuedReadSourceFileTasks: Record<string, number> = {};
-  const readSourceFileTasks: Promise<string[] | null>[] = [];
+  const filesToLines: Record<string, number[]> = {};
 
   if (contextLines > 0 && event.exception?.values) {
     for (const exception of event.exception.values) {
-      if (!exception.stacktrace?.frames) {
+      if (!exception.stacktrace?.frames?.length) {
         continue;
       }
 
-      // We want to iterate in reverse order as calling cache.get will bump the file in our LRU cache.
-      // This ends up prioritizes source context for frames at the top of the stack instead of the bottom.
+      // Maps preserve insertion order, so we iterate in reverse, starting at the
+      // outermost frame and closer to where the exception has occurred (poor mans priority)
       for (let i = exception.stacktrace.frames.length - 1; i >= 0; i--) {
         const frame = exception.stacktrace.frames[i];
-        // Call cache.get to bump the file to the top of the cache and ensure we have not already
-        // enqueued a read operation for this filename
-        if (frame.filename && !enqueuedReadSourceFileTasks[frame.filename] && !FILE_CONTENT_CACHE.get(frame.filename)) {
-          readSourceFileTasks.push(_readSourceFile(frame.filename));
-          enqueuedReadSourceFileTasks[frame.filename] = 1;
+
+        if (
+          typeof frame.filename !== 'string' ||
+          typeof frame.lineno !== 'number' ||
```

Comment:
I think a single `typeof frame.filename === `string` check is enough to ensure typesafety

Reply:
Is it not possible to have frames without a line? Here I'm just trying to ensure that the error has a lineno, else we can avoid reading the file

Reply:
Thanks for challenging me on this - let's keep the added safety and the double `typeof` check.

I originally posted this because I was fairly confident there should always be a `lineno` in node js environments, but who knows how our sdk is being used, it's not a good assumption to hold.